sqld/src/rpc/wal_log.rs

```diff
@@ -88,6 +89,12 @@ impl WalLog for WalLogService {
             let mut guard = self.replicas_with_hello.write().unwrap();
             guard.insert(replica_addr);
         }
-        Ok(tonic::Response::new(self.logger.hello_response().clone()))
+        let response = HelloResponse {
+            database_id: self.logger.database_id.to_string(),
+            generation_start_index: self.logger.generation.start_index,
+            generation_id: self.logger.generation.id.to_string(),
```

Comment:
Maybe we could store generation.id as string and avoid calling to_string every time? I know it's not performance critical but it's not that storing a string would be worse in any way.

Reply:
you'd have to clone it, so that does not make any difference

Reply:
It seems that converting number to string takes more work than just copying an array of bytes which works great with caches.

Reply:
maybe from a memcpy standpoint, but this is very likely trumped by the cost of allocation anyways. I'm deeply convinced that the performance gain from converting to a string here is insignificant compared to the gains of preserving type information

Reply:
The value of preserving type information is 0. No one uses this in form other than string.

It does not matter that some other cost trumps this. In well written systems there are no clear bottlenecks and everything takes 0.5-2% of performance. Then it becomes clear that small things add up to the slowness.

If there was a downside of storing this as string as oppose to UUID, it would be reasonable to ignore the performance but since there's absolutely no value in keeping this as UUID - I don't think it's worth keeping the value in suboptimal way.

Reply:
[A Uuid *is* a u128](https://en.wikipedia.org/wiki/Universally_unique_identifier), a string is only a representation of it. So when storing as a string we are storing a representation of a Uuid, not a u128. The representation can vary, but the Uuid it is representing is always the same. There are many reasons to use the Uuid type, over its string representation:
- it's Copy, whereas a string is not
- it lives on the stack, whereas a string is on the heap
- It's always comparable
- It's self-documenting
- It has a known size
- it comes with a full library to work with
- it is easily serializable in a zero-copy format
- it's typed

I have more arguments if this does not convince you :)

Reply:
> it's Copy, whereas a string is not

Do we ever copy/clone it? It seems that we put it into the meta object and it stays there for ever while being accessed through references, no?

>it lives on the stack, whereas a string is on the heap

Are you sure? https://docs.rs/uuid/latest/src/uuid/lib.rs.html#436

>It's always comparable

Are we ever comparing it on the primary?

> It's self-documenting

This is a fair point but using a `generation_uuid` name would be even more self documenting

> It has a known size

Are you sure? https://docs.rs/uuid/latest/src/uuid/lib.rs.html#436

> it comes with a full library to work with

Are we using any functions from that library?

>it is easily serializable in a zero-copy format

When we're serializing it to string it's both an allocation and a copy of the content, no?

>it's typed

What benefit does that give when we always pass it as string in the end?

Reply:
> Are you sure? https://docs.rs/uuid/latest/src/uuid/lib.rs.html#436
absolutely: https://docs.rs/uuid/latest/src/uuid/lib.rs.html#283

Let's look at string representation for a bit. There are at least 4 common and accepted ways to represent a Uuid
- braced: `{6aacd448-3de1-4e37-832c-6e48b8bb2ca0} `: https://docs.rs/uuid/latest/uuid/struct.Uuid.html#method.as_braced
- hyphenated: `6aacd448-3de1-4e37-832c-6e48b8bb2ca0`: https://docs.rs/uuid/latest/uuid/struct.Uuid.html#method.as_hyphenated
- simple: `6aacd4483de14e37832c6e48b8bb2ca0`: https://docs.rs/uuid/latest/uuid/struct.Uuid.html#method.as_simple
- urn: `urn:uuid:203e7fa8-65d6-46a6-ba25-72e7b222ed99 `: https://docs.rs/uuid/latest/uuid/struct.Uuid.html#method.as_urn
Now what that means is that if you want to compare them, you always need to parse them into a valid `Uuid` instance.

Let's say that a developer comes to the codebase. They are unaware of this optimization, and you see that the id is a string, how confident are you that they will go with the correct (and not obvious) way, that is convert back to Uuid, and compare them, or make sure that the representation is consistent with the one that was previously there, and that someone else relying on? What about other platforms? what about a uuid you receive from some other place? How do other platforms decide to encode it? Are you confident that such a mistake is caught in review?

We can also talk about perf, I have run some benches:

```
alloc string            time:   [20.232 ns 20.250 ns 20.272 ns]
                        change: [-0.5384% -0.3590% -0.1514%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 16 outliers among 100 measurements (16.00%)
  2 (2.00%) high mild
  14 (14.00%) high severe

Benchmarking to_string: Collecting 100 samples in estimated 5.0001 s (146M ite
                                                                              
to_string               time:   [33.970 ns 34.010 ns 34.069 ns]
                        change: [+0.6582% +0.8872% +1.1115%] (p = 0.00 < 0.05)
                        Change within noise threshold.
```

At first glance, It is true that converting to string is slower (34ns vs 20ns), but let's look a the fuller picture here, and consider what the actual cost/benefits are, for let's say 10000 replicas:
- assuming that the handshakes are serialized (of course they run concurrently), 10000*15ns = 150us in the lifetime of the cluster. After a full day we're looking at a (1.5e-8 / 86400) * 100 = 1.7361111e-11% improvement
- On the other hand, we get code that is a potential foot gun, and if the developer is careful enough, he will have to jump back and forth between the Uuid and its representation

Reply:
I don't agree with the developer argument. All the following questions are irrelevant in our situation:
```
What about other platforms? what about a uuid you receive from some other place? How do other platforms decide to encode it?
```
The field is set using Uuid so it's very easy to see what the format is. I don't expect anyone to try to parse such string without first looking at a place where the string is created :).

The second argument is valid though. Even though it's partially incorrect - it's 150us per primary restart not in the lifetime of the cluster. It's still very low so ok - let's keep it the way it is. I still don't see the reason to keep it as UUID and redo the work over and over again but since you proved the benefit is small - let it be.